crates/renderer/src/tree_renderer.rs

```diff
@@ -217,53 +250,164 @@ impl TreeRenderer {
     }
 
     #[allow(clippy::too_many_arguments)]
-    pub fn run_collect(
+    pub(crate) fn run_collect(
         &self,
         gpu: &Gpu,
+        world: &World,
         assets: &AssetCache,
         encoder: &mut wgpu::CommandEncoder,
         post_submit: &mut Vec<PostSubmitFunc>,
         resources_bind_group: &wgpu::BindGroup,
         entities_bind_group: &wgpu::BindGroup,
         collect_state: &mut RendererCollectState,
+        mesh_buffer: &MeshBuffer,
     ) {
-        tracing::debug!(
-            "Collecting tree renderer {:?} {:?}",
-            self.primitives.total_len(),
-            self.tree.keys().collect_vec()
-        );
-        let mut material_layouts = vec![UVec2::ZERO; self.material_indices.counter as usize];
+        let _span = tracing::debug_span!("collect", label = self.label).entered();
+        let mut material_layouts = vec![
+            MaterialLayout {
+                offset: 0,
+                primitive_count: 0
+            };
+            self.material_indices.counter as usize
+        ];
+
+        // Fill the material layouts with the offsets and primitive counts of the sub buffers
         for node in self.tree.values() {
             for mat in node.tree.values() {
                 let offset = self
                     .primitives
                     .buffer_offset(mat.primitives_subbuffer)
                     .unwrap();
-                material_layouts[mat.material_index as usize] =
-                    uvec2(offset as u32, mat.primitives.len() as u32);
+
+                material_layouts[mat.material_index as usize] = MaterialLayout {
+                    offset: offset.try_into().unwrap(),
+                    primitive_count: mat.primitives.len().try_into().unwrap(),
+                };
             }
         }
 
-        self.config.renderer_resources.collect.run(
-            gpu,
-            assets,
-            encoder,
-            post_submit,
-            resources_bind_group,
-            entities_bind_group,
-            &self.primitives,
-            collect_state,
-            self.primitives.total_len() as u32,
-            material_layouts,
-        );
+        assert_eq!(mem::size_of::<DrawIndexedIndirect>(), 20);
+
+        assert_eq!(mem::align_of::<DrawIndexedIndirect>(), 4);
+
+        if self.config.render_mode == RenderMode::Direct {
```

Comment:
Is this meant to be empty?

Reply:
Yes. The alternative is `if software_culling && render_mode == Indirect {} else if !software_culling && render_mode != Direct`

This makes for less inversions and duplicate boolean expressions